### PR TITLE
[minor] [feature]: Add MSIDBoundTokenProvider for in-process browser-native GetToken (POC)

### PR DESCRIPTION
**Pull request by agubuzomaximus**
## Proposed changes

Adds `MSIDBoundTokenProvider`, a Common Core seam that services a browser-native-message `GetToken` request **in-process** for a host such as OneAuth (embedded in Edge).

On unmanaged iOS the platform SSO Extension is unavailable, so the host cannot silently invoke the broker through `ASAuthorizationSingleSignOnProvider`. Instead the host hands the typed `MSIDBrowserNativeMessageGetTokenRequest` to this provider, which owns the orchestration that would otherwise live behind the SSO Extension (silent BART SPA redemption, or an interactive broker flip).

This PR is a **POC** that wires up the seam and proves the routing end-to-end; the real silent-redemption / interactive-broker-flip orchestration is layered on top of this provider in follow-up work.

I have code changes in OneAuth that prove functionality. This provider must be created to support OneAuth when SSO EXT is not an option. This is for iOS only not Mac OS. In Edge native app when Browser sends BrowserNativeMessagingRequest OneAuth will check if SSO Ext is available. if SSO EXT is absent then OneAuth will use the path of MSIDBoundTokenProvider via Common Core to pass the request in. This class will later hold real validation logic and routing for the response to be obtained either through silent or interactive flow. 

A Similar class exist in Broker (thats how this works with SSO ext), a later PR will be created to see how the broker code can be broken down since this class is now required in Common core. I have discussed this with Sergei before he went OOF. 

### What's included
- `IdentityCore/src/oauth2/token/MSIDBoundTokenProvider.{h,m}` — the provider. Validates the request (nil → `MSIDErrorInvalidInternalParameter`; blank `clientId`/`redirectUri` → `MSIDErrorInvalidDeveloperParameter`) and returns a serialized browser-native-message response payload (`transport: in_proc_common_core`).
- `IdentityCore/tests/MSIDBoundTokenProviderTests.m` — unit tests built from the real `MSIDBrowserNativeMessageGetTokenRequest` properties, covering the in-process success path and the missing-`clientId` validation path.

## Type of change

- [x] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [x] Small – No issues are expected. Additive, self-contained POC seam with no existing call sites; consumed only by the parallel OneAuth POC.

## Additional information

Paired with the OneAuth-side POC that routes `BrowserNativeMessagingGetTokenRequest` to this provider when the SSO Extension is unavailable. Branch: `maagubuzo/un_tb/poc/msidboundtokenprovider`, targeting `dev`.

### CODE REVIEW

File: IdentityCore/src/oauth2/token/MSIDBoundTokenProvider.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#import "MSIDRequestContext.h"
+
+@class MSIDBrowserNativeMessageGetTokenRequest;
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// Completion block for a bound-token acquisition.
+/// @param response Serialized browser-native-message response payload (JSON string) on success, otherwise nil.
+/// @param error Populated when acquisition fails, otherwise nil.
+typedef void (^MSIDBoundTokenProviderCompletionBlock)(NSString *_Nullable response, NSError *_Nullable error);
```
Comment (ameyapat) [inline IdentityCore/src/oauth2/token/MSIDBoundTokenProvider.h:35]: Can response be of type dictionary instead of string? Why does it need to be serialized to JSON?

Comment (agubuzomaximus) [reply IdentityCore/src/oauth2/token/MSIDBoundTokenProvider.h:35]: We serialize into a string because this is what one auth expects. So my code builds on top of Edge SSO into broker for BNM Get Token. Broker in the SSO path returns a string my goal is to keep that same return type otherwise One auth will have to convert to string in place of us. I believe browser wants a string of BNM and not a dictionary ... I believe that is the logic behind it. 

but to answer your question I turn to string to keep the same return type in one auth.

File: IdentityCore/src/oauth2/token/MSIDBoundTokenProvider.m [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+NSString *const MSID_BOUND_TOKEN_PROVIDER_LOG_PREFIX = @"[MSIDBoundTokenProvider]";
+
+@implementation MSIDBoundTokenProvider
+
+- (void)acquireBoundTokenWithRequest:(MSIDBrowserNativeMessageGetTokenRequest *)request
+                             context:(nullable id<MSIDRequestContext>)context
+                     completionBlock:(MSIDBoundTokenProviderCompletionBlock)completionBlock
+{
+    NSParameterAssert(completionBlock);
+    if (!completionBlock) return;
```
[the remark this answers is not in the document]
Comment (agubuzomaximus) [reply IdentityCore/src/oauth2/token/MSIDBoundTokenProvider.m:40]: this is handled in the next PR. I have also added proper logs on the completion and validation early returns

File: IdentityCore/src/oauth2/token/MSIDBoundTokenProvider.m [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    if ([NSString msidIsStringNilOrBlank:request.clientId] ||
+        [NSString msidIsStringNilOrBlank:request.redirectUri])
+    {
+        NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInvalidDeveloperParameter,
+                                         @"clientId and redirectUri are required to acquire a bound token.",
+                                         nil, nil, nil, context.correlationId, nil, NO);
+        completionBlock(nil, error);
+        return NO;
+    }
```
[the remark this answers is not in the document]
Comment (agubuzomaximus) [reply IdentityCore/src/oauth2/token/MSIDBoundTokenProvider.m:84]: this is a great comment it is acknowledged already in my second PR. there needs to be some work done to consolidate between the two as we do not want dup work in common core and broker.

**Comment by ameyapat** [comment]
Please increase code coverage via unit tests

**Review by ameyapat — CHANGES_REQUESTED** [review]

**Comment by agubuzomaximus** [comment]
Hi @ameyapat  I have added some test but this is a light change. please see this PR:
https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/pull/1875 it has more logic and heavy testing.

**Comment by ameyapat** [comment]
Consider adding unit tests for:

(1) add validation tests for nil (and optionally whitespace-only) clientId/redirectUri, not just empty strings
(2) add a payload-shape test for state == nil to assert state is omitted from the JSON
(3) add coverage for the completionBlock == nil guard path (even if assert behavior is build-config dependent)
(4) if practical, add a test for the JSON serialization fallback branch that returns "{}"

**Review by ameyapat — APPROVED** [review]
